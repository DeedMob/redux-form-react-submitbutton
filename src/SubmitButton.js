import React, { Component, PropTypes } from 'react';

export class SubmitButton extends Component {
  static defaultProps = {
    submitting: false,
    submitFailed: false,
    submitSucceeded: false,
    invalid: false,
    pristine: false,
    className: 'btn',
    styles: {},
    type: 'Submit',
    showIcons: true,
    timeout: 3000,
    labelSubmitting: '...Submitting',
    // start OK button messages
    labelSubmit: 'Submit',
    labelUpdate: 'Save Changes',
    labelPost: 'Post',
    labelCreate: 'Create',
    // start pristine button messages
    labelPristineSubmit: 'Incomplete',
    labelPristineUpdate: 'No Changes to Save',
    labelPristinePost: 'Incomplete',
    labelPristineCreate: 'Fill in required fields',
    // end pristine button messages
    labelInvalid: 'Invalid',
    labelSubmitFailed: 'Submission error',
    labelSubmitSucceeded: 'Success!'
  }
  static propTypes = {
    timeout: PropTypes.number,
    type: PropTypes.oneOf(['Create', 'Post', 'Update', 'Submit']),
    showIcons: PropTypes.bool,
    className: PropTypes.string,
    styles: PropTypes.object,
    submitting: PropTypes.bool,
    submitFailed: PropTypes.bool,
    submitSucceeded: PropTypes.bool,
    invalid: PropTypes.bool,
    pristine: PropTypes.bool,
    labelSubmitting: PropTypes.string,
    labelSubmit: PropTypes.string,
    labelUpdate: PropTypes.string,
    labelPost: PropTypes.string,
    labelCreate: PropTypes.string,
    labelPristineSubmit: PropTypes.string,
    labelPristineUpdate: PropTypes.string,
    labelPristinePost: PropTypes.string,
    labelPristineCreate: PropTypes.string,
    labelInvalid: PropTypes.string,
    labelSubmitFailed: PropTypes.string,
    labelSubmitSucceeded: PropTypes.string
  }
  constructor(props) {
    super(props);
    this.state = {
      lastActionWasSubmit: false,
      showSubmitState: false,
      hover: false,
    };
    this.onMouseOver = this.onMouseOver.bind(this);
    this.onMouseOut = this.onMouseOut.bind(this);
  }
  componentWillReceiveProps(nextProps) {
    if (nextProps.submitting) {
      this.setState({ lastActionWasSubmit: true });
    }
    if ((nextProps.submitFailed || nextProps.submitSucceeded) && this.state.lastActionWasSubmit) {
      this.setState({ showSubmitState: true, lastActionWasSubmit: false });
      setTimeout(() => this.setState({ showSubmitState: false }), this.props.timeout);
    }
  }
  onMouseOver() {
    this.setState({ hover: true });
  }
  onMouseOut() {
    this.setState({ hover: false });
  }
  render() {
    const { className, styles, showIcons } = this.props;
    const defaultLabel = this.props[`label${this.props.type}`];
    const defaultPristineLabel = this.props[`labelPristine${this.props.type}`];

    if (this.props.submitFailed && this.state.showSubmitState) {
      return (
        <button
          style={styles}
          className={`${className} btn-danger btn-outline`}
          disabled
          onMouseOver={this.onMouseOver}
          onMouseOut={this.onMouseOut}
        >
          {showIcons && <span><i className="fa fa-times" /></span>}
          {this.props.labelSubmitFailed}
        </button>
      );
    } else if (this.props.submitSucceeded && this.state.showSubmitState) {
      return (
        <button
          style={styles}
          className={`${className} btn-success btn-outline`}
          disabled
          onMouseOver={this.onMouseOver}
          onMouseOut={this.onMouseOut}
        >
          {showIcons && <span><i className="fa fa-check" /></span>}
          {this.props.labelSubmitSucceeded}
        </button>
      );
    } else if (this.props.submitting) {
      return (
        <button
          style={styles}
          className={`${className} btn-default btn-outline`}
          disabled
          onMouseOver={this.onMouseOver}
          onMouseOut={this.onMouseOut}
        >
          {showIcons && <span><i className="fa fa-spinner fa-pulse fa-fw" /></span>} {this.props.labelSubmitting}
        </button>
      );
    } else if (this.props.invalid && !this.props.pristine) {
      return (
        <button
          style={styles}
          className={`${className} btn-default btn-outline`}
          disabled
          onMouseOver={this.onMouseOver}
          onMouseOut={this.onMouseOut}
        >
          {this.state.hover ? this.props.labelInvalid : defaultLabel}
        </button>
      );
    } else if (!this.props.submitting && !this.props.pristine) {
      return (
        <button
          style={styles}
          className={`${className} btn-primary`}
          type="submit"
          onMouseOver={this.onMouseOver}
          onMouseOut={this.onMouseOut}
        >
          {showIcons && <span><i className="fa fa-paper-plane-o" /></span>}
          {defaultLabel}
        </button>
      );
    } else if (!this.props.submitting && this.props.pristine) {
      return (
        <button
          style={styles}
          className={`${className} btn-primary btn-outline`}
          disabled
          onMouseOver={this.onMouseOver}
          onMouseOut={this.onMouseOut}
        >
          {showIcons && <span><i className="fa fa-paper-plane-o" /></span>}
          {this.state.hover ? defaultPristineLabel : defaultLabel }
        </button>
      );
    }
    return (
      <button
        style={styles}
        className={`${className} btn-primary`}
        type="submit"
        onMouseOver={this.onMouseOver}
        onMouseOut={this.onMouseOut}
      >
        {showIcons && <span><i className="fa fa-paper-plane-o" /></span>}
        {defaultLabel}
      </button>
    );
  }
}

export default SubmitButton;
