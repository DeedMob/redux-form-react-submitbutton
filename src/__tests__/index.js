import React from 'react';
import { shallow, mount, render } from 'enzyme';
import renderer from 'react-test-renderer';

import { SubmitButton } from '../SubmitButton';


it('renders correctly', () => {
  const tree = renderer.create(<SubmitButton />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('Changes text on pristine hover', () => {
  const component = shallow(
    <SubmitButton submitting={false} pristine />
  );

  component.find('button').simulate('mouseover');
  expect(component.find('button').props()).toMatchSnapshot('mouseover');

  component.find('button').simulate('mouseout');
  expect(component.find('button').props()).toMatchSnapshot('mouseout');
});


it('to disable `button` on invalid', () => {
  const wrapper = shallow(<SubmitButton invalid />);
  expect(wrapper.find('button').props().disabled).toBeTruthy;
});

it('to not disable `button` if not invalid', () => {
  const wrapper = shallow(<SubmitButton invalid={false} />);
  expect(wrapper.find('button').props().disabled).toBeFalsy;
});

it('to add submitText', () => {
  const wrapper = shallow(<SubmitButton type="Submit" labelSubmit="something" />);
  expect(wrapper.find('button').text()).toMatch(/something/);
});

it('allows us to set props', () => {
  const wrapper = mount(<SubmitButton foo="bar" />);
  expect(wrapper.props().foo).toEqual('bar');
  wrapper.setProps({ bar: 'foo' });
  expect(wrapper.props().bar).toEqual('foo');
});

it('renders exactly one `button`', () => {
  const wrapper = render(<SubmitButton />);
  expect(wrapper.find('button').length).toEqual(1);
});
