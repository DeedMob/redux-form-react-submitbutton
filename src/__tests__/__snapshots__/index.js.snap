// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mouseout 1`] = `
Object {
  "children": Array [
    <span>
      <i
            className="fa fa-paper-plane-o"
      />
</span>,
    "Submit",
  ],
  "className": "btn btn-primary btn-outline",
  "disabled": true,
  "onMouseOut": [Function],
  "onMouseOver": [Function],
  "style": Object {},
}
`;

exports[`mouseover 1`] = `
Object {
  "children": Array [
    <span>
      <i
            className="fa fa-paper-plane-o"
      />
</span>,
    "Incomplete",
  ],
  "className": "btn btn-primary btn-outline",
  "disabled": true,
  "onMouseOut": [Function],
  "onMouseOver": [Function],
  "style": Object {},
}
`;

exports[`renders correctly 1`] = `
<button
  className="btn btn-primary"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  style={Object {}}
  type="submit"
>
  <span>
    <i
      className="fa fa-paper-plane-o"
    />
  </span>
  Submit
</button>
`;
